### PR TITLE
feat(protocol): improve guardian prover

### PR DESCRIPTION
**Pull request by dantaik**
## Change Summary

- Changed from requiring a majority (over half) to a fixed minimum. This increases flexibility in approval processes, **we need this feature**.
- Supports off-chain signature aggregation for future gas optimization in guardian prover infra.
- Merged Guardians and GuardianProver into one.

### CODE REVIEW

File: packages/protocol/contracts/L1/provers/GuardianProver.sol [hunk 1]
```diff
@@ -44,13 +127,94 @@ contract GuardianProver is Guardians {
...
+        if (_signatures.length < minGuardians) revert INVALID_SIGNATURES();
+
+        bytes32 hash = keccak256(abi.encode("APPROVE", _meta, _tran));
+        address lastGuardian;
+
+        for (uint256 i; i < minGuardians; ++i) {
+            address guardian = ECDSA.recover(hash, _signatures[i]);
+
+            // This makes sure there are not duplicate addresses
+            if (uint160(guardian) <= uint160(lastGuardian) || guardianIds[guardian] == 0) {
```
[the remark this answers is not in the document]
Comment (dantaik) [reply packages/protocol/contracts/L1/provers/GuardianProver.sol:165]: should be "<"

[the remark this answers is not in the document]
Comment (dantaik) [reply packages/protocol/contracts/L1/provers/GuardianProver.sol:165]: if the two addresses are the same, then the function shall revert.

[the remark this answers is not in the document]
Comment (adaki2004) [reply packages/protocol/contracts/L1/provers/GuardianProver.sol:165]: ah yeah, i misread and thought it is a pass-through filter. 🖖

**Review by davidtaikocha — APPROVED** [review]

**Comment by Brechtpd** [comment]
Why not make this contract the `GuardianVerifier` directly? Because currently the `GuardianVerifier` is kind of a strange one where the `msg.sender` is checked which makes it work in a different way than any other verifiers that we'll have. It also makes it impossible (at least in a reasonable way) to combine with other proof types because they would also have to pass through this contract if a guardian proof is required next to the other proofs.

Instead, if this contract is the `GuardianVerifier`, it can easily be combined with other proof types because `verifyProof` can just check the signatures directly or check if the block is approved using `isApproved` using the onchain way of approving blocks.

Alternatively  `GuardianVerifier` could also call back into `resolve("guardian_prover", false)` using a `isSignatureValid` kind of method to still keep the two contract separate, but I don't immediately see much benefit of doing it like that.

**Comment by otara81** [comment]
Wen

**Comment by otara81** [comment]
Wen

**Pull request closed without merging**